src/bubble_swarm/utils.rs

```diff
@@ -0,0 +1,75 @@
+use crate::math2d::{circle::Circle, point::Point, regioned_angle::RegionedAngle, vector::Vector};
+
+use super::{angle_area::AngleArea, source_circle::SourceCircle};
+
+pub fn get_mid_point(from: Point, to: Point) -> Point {
+    from + Vector::from((from, to)).scale(0.5)
+}
+
+pub enum CircleIntersection {
+    None,
+    One(Point),
+    Two(Point, Point),
+}
+
+pub fn calculate_circle_intersection(first: &Circle, second: &Circle) -> CircleIntersection {
+    // Calculate the distance between the centers.
+    let distance = Vector::from((first.get_center(), second.get_center())).magnitude();
+
+    // Check if the circles are completely separate.
+    if distance > first.get_radius() + second.get_radius() {
+        return CircleIntersection::None;
+    }
+
+    // Check if one circle is contained within the other.
+    if distance < (first.get_radius() - second.get_radius()).abs() {
+        return CircleIntersection::None;
+    }
+
+    // Check if the circles are touching.
+    if distance == first.get_radius() + second.get_radius() {
```

Comment:
I don't know if get_radius() returns a float here. If not that is bad. If yes it is more robust to do (first.get_radius() - second.get_radius()).ans() < 1e-10.

Reply:
You mean not to check the equality to distance but like 

`if (first.get_radius() + second.get_radius()).abs() < 1e-10 {
        return CircleIntersection::One(get_mid_point(first.get_center(), second.get_center()));
    }`